climada/util/calibrate/impact_func.py

```diff
@@ -0,0 +1,225 @@
+"""Module for calibrating impact functions"""
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field, InitVar
+from typing import Callable, Mapping, Optional, Tuple, Union, Any, Dict, List
+from numbers import Number
+
+import numpy as np
+import pandas as pd
+from scipy.optimize import (
+    Bounds,
+    LinearConstraint,
+    NonlinearConstraint,
+    minimize,
+)
+from bayes_opt import BayesianOptimization
+
+from climada.hazard import Hazard
+from climada.entity import Exposures, ImpactFunc, ImpactFuncSet
+from climada.engine import Impact, ImpactCalc
+
+
+def cost_func_rmse(
+    impact: Impact,
+    data: pd.DataFrame,
+    impact_proc: Callable[[Impact], pd.DataFrame] = lambda x: x.impact_at_reg(),
+) -> Number:
+    return np.sqrt(np.mean(((impact_proc(impact) - data) ** 2).to_numpy()))
+
+
+def impf_step_generator(threshold: Number, paa: Number) -> ImpactFuncSet:
+    return ImpactFuncSet(
+        [
+            ImpactFunc.from_step_impf(
+                haz_type="RF", intensity=(0, threshold, 100), paa=(0, paa)
+            )
+        ]
+    )
+
+
+ConstraintType = Union[LinearConstraint, NonlinearConstraint, Mapping]
+
+
+@dataclass
+class Input:
+    """Define the static input for a calibration task
+
+    Parameters
+    ----------
+    hazard : climada.Hazard
+        Hazard object to compute impacts from
+    exposure : climada.Exposures
+        Exposures object to compute impacts from
+    data : pandas.Dataframe
+        The data to compare computed impacts to. Index: Event IDs matching the IDs of
+        ``hazard``. Columns: Arbitrary columns.
+    cost_func : Callable
+        Function that takes an ``Impact`` object and a ``pandas.Dataframe`` as argument
+        and returns a single number. The optimization algorithm will try to minimize this
+        number. See this module for a suggestion of cost functions.
+    impact_func_gen : Callable
```

Comment:
I would avoid the `_gen` as this can easily be confused for a `generator` in Python.

Reply:
How would you call a function that returns a function?

Reply:
I would try for a name that helps me understand what happens, not what it technically is. In this case, I think `impact_function` would be enough, as this is what it is. Or maybe `parametrized_impact_function`.

Reply:
Careful, it's not am impact function, it is a function that *returns* an impact function. So `impact_function` would be severely misleading

Reply:
Well, it is a function. It is true that it might be confused with the object `ImpactFunction`. But it is definitely simply a parametrized function.

Reply:
The attribute is now called `impact_func_creator`